### PR TITLE
Update x86_64 and arm64 bindings to v5.13

### PR DESCRIPTION
**Pull request by sboeuf**
This patch is a proposal to simplify kvm-bindings crate, moving to the
latest Linux version 5.13.

First thing, instead of maintaining multiple versions (4.14, 4.20 and
now 5.13), this patch only keeps 5.13, and moving forward, we would only
keep the most up to date version.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

### CODE REVIEW

**Comment by sboeuf** [comment]
@andreeaflorescu I know I'm removing lots of things through this PR, but I thought I'd give it a try to see what y'all think about it.

**Comment by andreeaflorescu** [comment]
I was thinking we can still keep the arm & x86 bindings. I am a bit unsure who is using this crate, I saw some other dependencies on it from project I was not aware of before. Is it too much trouble to keep them?

Can you also test this with kvm-ioctls to be sure it still works? We can open an RFC PR in kvm-ioctls to patch kvm-bindings to use your fork and do a dry run of the tests.

We also need to check that this still works in the Firecracker use case, but I would assume it does.

I think it would be better to split the PR in multiple commits:
- one that removes the features, and first just keeps v4.20
- and one that does the update from v4.20 to v5.13
It should be clear that way what are the changes.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
 ### Bindgen
 The bindings are currently generated using
-[bindgen](https://crates.io/crates/bindgen) version 0.46.0:
```
Comment (andreeaflorescu) [inline CONTRIBUTING.md:7]: We should keep this information even if we recommend using the latest available bindings to generate new structures. It is good for reproducing the binaries with the exact version.

Can you add the version you used for generating the bindings?

**Comment by sboeuf** [comment]
>I was thinking we can still keep the arm & x86 bindings. I am a bit unsure who is using this crate, I saw some other dependencies on it from project I was not aware of before. Is it too much trouble to keep them?

No worries I can keep them. That being said, I couldn't find `kvm.h` in the headers generated along with the kernel for `arm` arch. Should I simply use the one from `arm64`? @MrXinWang could you give some input there?

>Can you also test this with kvm-ioctls to be sure it still works? We can open an RFC PR in kvm-ioctls to patch kvm-bindings to use your fork and do a dry run of the tests.

Yes I can open a draft PR in kvm-ioctls based on my fork.

> We also need to check that this still works in the Firecracker use case, but I would assume it does.

Can you try this out on your side?

>I think it would be better to split the PR in multiple commits:
>
>   - one that removes the features, and first just keeps v4.20
>   - and one that does the update from v4.20 to v5.13
>    It should be clear that way what are the changes.

Yes that sounds like a good plan :)

File: CONTRIBUTING.md (hunk 1 above)
Comment (sboeuf) [reply CONTRIBUTING.md:7]: Sure!

**Comment by MrXinWang** [comment]
> > I was thinking we can still keep the arm & x86 bindings. I am a bit unsure who is using this crate, I saw some other dependencies on it from project I was not aware of before. Is it too much trouble to keep them?
> 
> No worries I can keep them. That being said, I couldn't find `kvm.h` in the headers generated along with the kernel for `arm` arch. Should I simply use the one from `arm64`? @MrXinWang could you give some input there?

Hi both, I spent some time digging into the patchwork, and found that the arm32 hypervisor support in kvm has been removed - It explains why @sboeuf cannot found the `kvm.h` for arm32 . See https://patchwork.kernel.org/project/kvm/cover/20200210141324.21090-1-maz@kernel.org/. Therefore I think yes?

**Comment by sboeuf** [comment]
@MrXinWang thanks for checking. So I'll do the same thing we do for x86 (which means we use the same bindings for both 32 and 64 bits).

**Comment by MrXinWang** [comment]
As I am not familiar with x86 system (sorry...). Maybe for the x86 system @andreeaflorescu can give some input or suggestions :)) Thanks!

**Comment by sboeuf** [comment]
@andreeaflorescu https://github.com/rust-vmm/kvm-bindings/pull/43 is the first step (cleaning up 4.14).

**Comment by andreeaflorescu** [comment]
I'll look into the changes, but I am going on vacation starting with Monday, and I have a few other things I need to get done by then. I'll let you know about my progress.

My plan is to:
- do a diff to see what's changed between 4.20 and 5.13 in terms of structures (and other binding specific defines)
- update the Firecracker's bindings
- run tests to see if it works with the current bindings
- check if upgrading from one version to another works

**Comment by sboeuf** [comment]
@andreeaflorescu ok this PR is up to date now, based on the previous discussions. Take your time to test it on Firecracker, and once you can confirm it is working fine, we can move forward and merge it (when you're back from your vacation).
In the meantime I'll send a draft PR on kvm-ioctls relying on this branch.

**Comment by sboeuf** [comment]
Here is the `kvm-ioctls` draft: https://github.com/rust-vmm/kvm-ioctls/pull/160

**Review by andreeaflorescu — DISMISSED** [review]

**Comment by sboeuf** [comment]
@andreeaflorescu if you agree I can simply close #43 and we can merge (when ready) this one directly since it includes #43.

**Comment by andreeaflorescu** [comment]
Ah, I mistakenly approved this one, I meant to approve the other one. I didn't yet got to integrate the changes in Firecracker. I just did a read up on the diffs. It looks like some structures we're using in Firecracker are changing state (like the kvm_irqchip), but I believe they can be easily addressed.

@alexandruag said he'll take a look at this one as well while I am off.

**Comment by andreeaflorescu** [comment]
> @andreeaflorescu if you agree I can simply close #43 and we can merge (when ready) this one directly since it includes #43.

We can also do as you propose.

**Comment by andreeaflorescu** [comment]
@sboeuf what bindgen version did you use for generating the new bindings? I think we might need to update this in the documentation as well. Locally it looks like for me it installs v0.59.0.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -35,22 +35,22 @@ cd ~
...
 # This will generate the headers for the targeted architecture and place them
 # in the user specified directory. In this case, we generate them in the
-# arm64_v4_20_headers directory.
-make headers_install ARCH=arm64 INSTALL_HDR_PATH=arm64_v4_20_headers
-cd arm64_v4_20_headers
-bindgen include/linux/kvm.h -o bindings_v4_20_0.rs \
+# arm64_v5_13_headers directory.
+make headers_install ARCH=arm64 INSTALL_HDR_PATH=arm64_v5_13_headers
+cd arm64_v5_13_headers
+bindgen include/linux/kvm.h -o bindings_v5_13_0.rs \
```
Comment (andreeaflorescu) [inline CONTRIBUTING.md:46]: This was a bit wrong before, there are some `\` missing:
```
bindgen include/linux/kvm.h -o bindings_v5_13_0.rs \
  --with-derive-default \
  --with-derive-partialeq \
  -- -Iinclude
```

File: src/arm64/bindings.rs [hunk 3]
```diff
@@ -826,7156 +844,8500 @@ pub type __sum16 = __u16;
...
             "Offset of field: ",
-            stringify!(kvm_s390_ext_info),
+            stringify!(kvm_guest_debug),
             "::",
             stringify!(pad)
         )
     );
     assert_eq!(
-        unsafe { &(*(::std::ptr::null::<kvm_s390_ext_info>())).ext_params2 as *const _ as usize },
+        unsafe { &(*(::std::ptr::null::<kvm_guest_debug>())).arch as *const _ as usize },
```
Comment (andreeaflorescu) [inline src/arm64/bindings.rs:8490]: If we're using the newer bindgen (v0.59.0) this is not using `zeroed` anymore:
```rust
impl Default for kvm_xen_hvm_attr {
    fn default() -> Self {
        let mut s = ::std::mem::MaybeUninit::<Self>::uninit();
        unsafe {
            ::std::ptr::write_bytes(s.as_mut_ptr(), 0, 1);
            s.assume_init()
        }
    }
}
```

`MaybeUninit.assume_init()` is the preferred way to zero a structure.

Comment (andreeaflorescu) [inline src/arm64/bindings.rs:6554]: With the latest bindgen version, this structure also derives Debug and PartialEq. This happens for a few other structures as well.

**Review by andreeaflorescu — COMMENTED** [review]
I did a re-run with the latest version of bindgen. My comments reflect the changes. I think it might be worth it to use the latest version to generate the bindings as there are quite a few more structures that can derive partialeq & debug and those are helpful when writing tests/debugging.

I also finished my analysis of changes on both x86_64 and aarch64, and I'll publish it (still need to figure out where/how :laughing:)

**Comment by sboeuf** [comment]
@andreeaflorescu sounds good, I'll update the PR with bindings generated from `bindgen` `v0.59.0`

**Comment by sboeuf** [comment]
@andreeaflorescu PR updated! Please have a look.

**Comment by AlexandruCihodaru** [comment]
Hello, I have tested it with Firecracker, so far all integration tests are passing.

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -16,58 +16,10 @@ steps:
           image: "rustvmm/dev:v4"
```
Comment (andreeaflorescu) [inline .buildkite/pipeline.yml:16]: As we discussed offline, we need to update this to v12 (same for all the other steps).

**Review by andreeaflorescu — COMMENTED** [review]
The coverage also needs to be updated.

Once that is fixed, I think we can merge the PR.

**Comment by andreeaflorescu** [comment]
> Hello, I have tested it with Firecracker, so far all integration tests are passing.

Thanks, @AlexandruCihodaru!

**Comment by sboeuf** [comment]
@andreeaflorescu PR updated accordingly :)

**Review by andreeaflorescu — APPROVED** [review]

**Comment by sboeuf** [comment]
@andreeaflorescu can we find someone else to approve and merge this PR?

**Review by acatangiu — APPROVED** [review]
Changes look good!

**Review by alexandruag — APPROVED** [review]
Looks good, apologies for the delay.

**Pull request merged**